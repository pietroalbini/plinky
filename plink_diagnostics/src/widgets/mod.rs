mod table;

pub use self::table::Table;
