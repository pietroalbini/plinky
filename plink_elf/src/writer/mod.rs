mod cursor;
mod layout;

pub(crate) use self::cursor::WriteCursor;
pub(crate) use self::layout::WriteLayoutError;

use crate::errors::WriteError;
use crate::ids::{ElfIds, StringIdGetters};
use crate::raw::{
    RawHeader, RawIdentification, RawPadding, RawProgramHeader, RawSectionHeader, RawType,
};
use crate::utils::WriteSeek;
use crate::writer::layout::{Part, WriteLayout};
use crate::{
    ElfABI, ElfClass, ElfEndian, ElfMachine, ElfObject, ElfSectionContent, ElfSegmentContent,
    ElfSegmentType, ElfType,
};
use std::collections::BTreeMap;

pub(crate) struct Writer<'a, I>
where
    I: ElfIds,
    I::StringId: StringIdGetters<I>,
{
    cursor: WriteCursor<'a>,
    layout: WriteLayout<I>,
    object: &'a ElfObject<I>,
}

impl<'a, I> Writer<'a, I>
where
    I: ElfIds,
    I::StringId: StringIdGetters<I>,
{
    pub(crate) fn new(
        write_to: &'a mut dyn WriteSeek,
        object: &'a ElfObject<I>,
    ) -> Result<Self, WriteError> {
        Ok(Self {
            cursor: WriteCursor::new(write_to, object),
            layout: WriteLayout::new(object)?,
            object,
        })
    }

    pub(crate) fn write(mut self) -> Result<(), WriteError> {
        let parts = self.layout.parts().iter().cloned().collect::<Vec<_>>();
        for part in &parts {
            match part {
                Part::Identification => self.write_identification()?,
                Part::Header => self.write_header()?,
                Part::SectionHeaders => self.write_section_headers()?,
                Part::ProgramHeaders => self.write_program_headers()?,
                Part::ProgramSection(id) => self.write_program_section(id)?,
                Part::StringTable(id) => self.write_string_table(id)?,
                Part::Padding(_) => self.write_padding(part)?,
            }
        }
        Ok(())
    }

    fn write_identification(&mut self) -> Result<(), WriteError> {
        let identification = RawIdentification {
            magic: [0x7F, b'E', b'L', b'F'],
            class: match self.object.env.class {
                ElfClass::Elf32 => 1,
                ElfClass::Elf64 => 2,
            },
            endian: match self.object.env.endian {
                ElfEndian::Little => 1,
            },
            version: 1,
            abi: match self.object.env.abi {
                ElfABI::SystemV => 0,
            },
            abi_version: match self.object.env.abi {
                ElfABI::SystemV => 0,
            },
            padding: RawPadding,
        };
        identification.write(&mut self.cursor)
    }

    fn write_header(&mut self) -> Result<(), WriteError> {
        let header = RawHeader {
            type_: match self.object.type_ {
                ElfType::Relocatable => 1,
                ElfType::Executable => 2,
                ElfType::SharedObject => 3,
                ElfType::Core => 4,
            },
            machine: match self.object.env.machine {
                ElfMachine::X86 => 3,
                ElfMachine::X86_64 => 62,
            },
            version: 1,
            entry: self.object.entry.map(|n| n.get()).unwrap_or(0),
            program_headers_offset: self.layout.metadata(&Part::ProgramHeaders).offset,
            section_headers_offset: self.layout.metadata(&Part::SectionHeaders).offset,
            flags: 0,
            elf_header_size: self.raw_type_size::<RawIdentification>()
                + self.raw_type_size::<RawHeader>(),
            program_header_size: self.raw_type_size::<RawProgramHeader>(),
            program_header_count: self.object.segments.len() as _,
            section_header_size: self.raw_type_size::<RawSectionHeader>(),
            section_header_count: self.object.sections.len() as _,
            section_names_table_index: self.find_section_names_string_table()?,
        };
        header.write(&mut self.cursor)
    }

    fn write_section_headers(&mut self) -> Result<(), WriteError> {
        for (id, section) in &self.object.sections {
            if let ElfSectionContent::Null = section.content {
                RawSectionHeader::zero().write(&mut self.cursor)?;
                continue;
            }

            let metadata = self.layout.metadata_of_section(id);
            let header = RawSectionHeader {
                name_offset: section.name.offset(),
                type_: match &section.content {
                    ElfSectionContent::Null => unreachable!(),
                    ElfSectionContent::Program(_) => 1,
                    ElfSectionContent::SymbolTable(_) => todo!(),
                    ElfSectionContent::StringTable(_) => 3,
                    ElfSectionContent::RelocationsTable(_) => todo!(),
                    ElfSectionContent::Note(_) => todo!(),
                    ElfSectionContent::Unknown(_) => panic!("unknown section"),
                },
                flags: match &section.content {
                    ElfSectionContent::Program(program) => {
                        (program.perms.write as u64)
                            | (program.perms.read as u64) << 1
                            | (program.perms.execute as u64) << 2
                    }
                    _ => 0,
                },
                memory_address: section.memory_address,
                offset: metadata.offset,
                size: metadata.len,
                link: 0,
                info: 0,
                addr_align: 0x1,
                entries_size: 0,
            };
            header.write(&mut self.cursor)?;
        }
        Ok(())
    }

    fn write_program_headers(&mut self) -> Result<(), WriteError> {
        for segment in &self.object.segments {
            let (metadata, section) = match segment.content.as_slice() {
                [ElfSegmentContent::Section(id)] => (
                    self.layout.metadata_of_section(id),
                    self.object.sections.get(id).unwrap(),
                ),
                [ElfSegmentContent::Unknown(_)] => todo!(),
                _ => todo!(),
            };

            let header = RawProgramHeader {
                type_: match segment.type_ {
                    ElfSegmentType::Null => 0,
                    ElfSegmentType::Load => 1,
                    ElfSegmentType::Dynamic => 2,
                    ElfSegmentType::Interpreter => 3,
                    ElfSegmentType::Note => 4,
                    ElfSegmentType::ProgramHeaderTable => 5,
                    ElfSegmentType::Unknown(_) => panic!("unknown segment"),
                },
                file_offset: metadata.offset,
                virtual_address: section.memory_address,
                reserved: 0,
                file_size: metadata.len,
                memory_size: metadata.len,
                flags: (segment.perms.execute as u32)
                    | (segment.perms.write as u32) << 1
                    | (segment.perms.read as u32) << 2,
                align: 0x1000,
            };
            header.write(&mut self.cursor)?;
        }
        Ok(())
    }

    fn write_string_table(&mut self, id: &I::SectionId) -> Result<(), WriteError> {
        let ElfSectionContent::StringTable(table) = &self.object.sections.get(id).unwrap().content
        else {
            panic!("section {id:?} is not a string table");
        };

        for string in table.all() {
            self.cursor.write_bytes(string.as_bytes())?;
            self.cursor.write_bytes(b"\0")?;
        }
        Ok(())
    }

    fn write_program_section(&mut self, id: &I::SectionId) -> Result<(), WriteError> {
        let ElfSectionContent::Program(program) = &self.object.sections.get(id).unwrap().content
        else {
            panic!("section {id:?} is not a program section");
        };
        self.cursor.write_bytes(&program.raw.0)
    }

    fn write_padding(&mut self, part: &Part<I::SectionId>) -> Result<(), WriteError> {
        let metadata = self.layout.metadata(part);
        let padding = vec![0; metadata.len as usize];
        self.cursor.write_bytes(&padding)
    }

    fn find_section_names_string_table(&self) -> Result<u16, WriteError> {
        let section_ids_to_indices = self
            .object
            .sections
            .keys()
            .enumerate()
            .map(|(idx, id)| (id.clone(), idx as u16))
            .collect::<BTreeMap<_, _>>();

        let mut string_table_section_id = None;
        for section in self.object.sections.values() {
            match &string_table_section_id {
                Some(existing_id) if section.name.section() == existing_id => {}
                Some(_) => return Err(WriteError::InconsistentSectionNamesTableId),
                None => string_table_section_id = Some(section.name.section().clone()),
            }
        }

        string_table_section_id
            .and_then(|id| section_ids_to_indices.get(&id))
            .copied()
            .ok_or(WriteError::MissingSectionNamesTable)
    }

    fn raw_type_size<T: RawType>(&self) -> u16 {
        T::size(self.object.env.class) as _
    }
}
