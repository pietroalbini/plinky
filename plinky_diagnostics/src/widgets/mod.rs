mod group;
mod hex_dump;
mod quoted_text;
mod table;
mod text;

pub use self::group::WidgetGroup;
pub use self::hex_dump::HexDump;
pub use self::quoted_text::QuotedText;
pub use self::table::Table;
pub use self::text::Text;
use crate::WidgetWriter;

pub trait Widget {
    fn render(&self, writer: &mut WidgetWriter<'_>);

    fn render_to_string(&self) -> String {
        let mut buffer = String::new();
        let mut writer = WidgetWriter::new(&mut buffer);
        self.render(&mut writer);
        buffer
    }
}

impl Widget for Box<dyn Widget> {
    fn render(&self, writer: &mut WidgetWriter<'_>) {
        self.as_ref().render(writer)
    }
}
