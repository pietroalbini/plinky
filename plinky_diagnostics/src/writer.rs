pub struct WidgetWriter<'a> {
    buffer: &'a mut String,

    indent: Vec<Indent>,
    last_char: Option<char>,
}

impl<'a> WidgetWriter<'a> {
    pub fn new(buffer: &'a mut String) -> Self {
        Self { buffer, indent: Vec::new(), last_char: None }
    }

    pub(crate) fn push(&mut self, content: char) {
        let indents_to_consider = if content == '\n' {
            indents_for_empty_line(&mut self.indent)
        } else {
            &mut self.indent
        };
        for indent in indents_to_consider {
            if !indent.printed_this_line {
                self.buffer.push_str(&indent.content);
                indent.printed_this_line = true;
            }
        }

        self.buffer.push(content);
        self.last_char = Some(content);

        if content == '\n' {
            for indent in &mut self.indent {
                indent.printed_this_line = false;
            }
        }
    }

    pub(crate) fn push_str(&mut self, content: &str) {
        for chr in content.chars() {
            self.push(chr);
        }
    }

    pub(crate) fn push_indent(&mut self, indent: &str, mode: IndentMode) {
        self.indent.push(Indent::new(indent, mode));
    }

    pub(crate) fn pop_indent(&mut self) {
        self.indent.pop();
    }

    pub(crate) fn last_char(&self) -> Option<char> {
        self.last_char
    }
}

fn indents_for_empty_line(indents: &mut [Indent]) -> &mut [Indent] {
    let skip = indents
        .iter()
        .rev()
        .take_while(|indent| matches!(indent.mode, IndentMode::HideOnEmptyLines))
        .count();
    let len = indents.len();
    &mut indents[..len - skip]
}

#[derive(Debug, PartialEq, Eq)]
struct Indent {
    content: String,
    mode: IndentMode,
    printed_this_line: bool,
}

impl Indent {
    fn new(content: &str, mode: IndentMode) -> Self {
        Self { content: content.to_string(), mode, printed_this_line: false }
    }
}

#[derive(Debug, PartialEq, Eq, Clone, Copy)]
pub(crate) enum IndentMode {
    ShowAlways,
    HideOnEmptyLines,
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_indents_for_empty_line() {
        let i = |mode| Indent::new("_", mode);

        assert_eq!(&mut [] as &mut [Indent], indents_for_empty_line(&mut []));
        assert_eq!(
            &mut [] as &mut [Indent],
            indents_for_empty_line(&mut [i(IndentMode::HideOnEmptyLines)])
        );
        assert_eq!(
            &mut [i(IndentMode::ShowAlways)],
            indents_for_empty_line(&mut [i(IndentMode::ShowAlways)])
        );
        assert_eq!(
            &mut [i(IndentMode::ShowAlways)],
            indents_for_empty_line(&mut [
                i(IndentMode::ShowAlways),
                i(IndentMode::HideOnEmptyLines)
            ])
        );
        assert_eq!(
            &mut [
                i(IndentMode::ShowAlways),
                i(IndentMode::HideOnEmptyLines),
                i(IndentMode::HideOnEmptyLines),
                i(IndentMode::ShowAlways),
            ],
            indents_for_empty_line(&mut [
                i(IndentMode::ShowAlways),
                i(IndentMode::HideOnEmptyLines),
                i(IndentMode::HideOnEmptyLines),
                i(IndentMode::ShowAlways),
            ])
        );
        assert_eq!(
            &mut [
                i(IndentMode::ShowAlways),
                i(IndentMode::HideOnEmptyLines),
                i(IndentMode::HideOnEmptyLines),
                i(IndentMode::ShowAlways),
            ],
            indents_for_empty_line(&mut [
                i(IndentMode::ShowAlways),
                i(IndentMode::HideOnEmptyLines),
                i(IndentMode::HideOnEmptyLines),
                i(IndentMode::ShowAlways),
                i(IndentMode::HideOnEmptyLines),
                i(IndentMode::HideOnEmptyLines),
            ])
        );
    }
}
