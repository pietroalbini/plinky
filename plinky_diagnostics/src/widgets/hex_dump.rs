use crate::WidgetWriter;
use crate::widgets::{Table, Widget};
use std::fmt::Write;

const BYTES_PER_LINE: usize = 16;

pub struct HexDump {
    data: Vec<u8>,
}

impl HexDump {
    pub fn new(data: impl Into<Vec<u8>>) -> Self {
        Self { data: data.into() }
    }
}

impl Widget for HexDump {
    fn render(&self, writer: &mut WidgetWriter<'_>) {
        let mut hex = String::new();
        let mut ascii = String::new();

        for chunk in self.data.chunks(BYTES_PER_LINE) {
            for (i, byte) in chunk.iter().copied().enumerate() {
                if i > 0 {
                    hex.push(' ');
                }
                hex.write_fmt(format_args!("{byte:0>2x}")).unwrap();
                if show_as_ascii(byte) {
                    ascii.push(byte as char);
                } else {
                    ascii.push('.');
                }
            }
            hex.push('\n');
            ascii.push('\n');
        }

        let mut table = Table::new();
        table.add_row([hex.trim(), ascii.trim()]);
        table.render(writer);
    }
}

fn show_as_ascii(byte: u8) -> bool {
    byte.is_ascii_alphanumeric() || byte.is_ascii_punctuation() || byte == b' '
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::configure_insta;
    use insta::assert_snapshot;

    #[test]
    fn test_hex_dump_hello() {
        let _config = configure_insta();

        assert_snapshot!(HexDump::new(b"Hello world").render_to_string());
    }

    #[test]
    fn test_hex_dump_256() {
        let _config = configure_insta();

        let data = (0u8..=255u8).collect::<Vec<_>>();
        assert_snapshot!(HexDump::new(data).render_to_string());
    }
}
