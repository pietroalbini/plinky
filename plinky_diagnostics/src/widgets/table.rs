use crate::widgets::Widget;
use crate::WidgetWriter;

const UNICODE_CHARSET: TableCharset = TableCharset {
    vertical_separator: '│',
    horizontal_separator: '─',
    first_junction: TableJunctionCharset { first: '╭', middle: '┬', last: '╮' },
    middle_junction: TableJunctionCharset { first: '├', middle: '┼', last: '┤' },
    last_junction: TableJunctionCharset { first: '╰', middle: '┴', last: '╯' },
};

pub struct Table {
    charset: &'static TableCharset,
    title: Option<String>,
    state: TableState,
}

impl Table {
    pub fn new() -> Self {
        Self { charset: &UNICODE_CHARSET, title: None, state: TableState::Empty }
    }

    pub fn set_title(&mut self, title: impl Into<String>) {
        self.title = Some(title.into());
    }

    pub fn add_row<I, V>(&mut self, row: I)
    where
        V: Into<String>,
        I: IntoIterator<Item = V>,
    {
        let mut count = 0;
        let row = row
            .into_iter()
            .map(|cell| cell.into().split('\n').map(|s| s.to_string()).collect::<Vec<_>>())
            .inspect(|_| count += 1);

        match &mut self.state {
            TableState::Empty => {
                let mut cells_len = Vec::new();
                self.state = TableState::HasContent {
                    content: row
                        .inspect(|cell| {
                            cells_len.push(cell.iter().map(|l| l.len()).max().unwrap_or(0))
                        })
                        .collect(),
                    cells_len,
                    cells_count: count,
                };
            }
            TableState::HasContent { cells_count, cells_len, content } => {
                content.extend(row.enumerate().map(|(pos, cell)| {
                    let cell_len = cell.iter().map(|c| c.len()).max().unwrap_or(0);
                    match cells_len.get_mut(pos) {
                        Some(len) => *len = (*len).max(cell_len),
                        None => cells_len.push(cell_len),
                    }
                    cell
                }));
                if count != *cells_count {
                    panic!("other rows have {cells_count} cells, while this row has {count} cells");
                }
            }
        }
    }

    fn render_horizontal_border(
        &self,
        writer: &mut WidgetWriter,
        cells_len: &[usize],
        junction: &TableJunctionCharset,
    ) {
        writer.push(junction.first);
        for (idx, len) in cells_len.iter().enumerate() {
            if idx != 0 {
                writer.push(junction.middle);
            }
            for _ in 0..(*len + 2) {
                writer.push(self.charset.horizontal_separator);
            }
        }
        writer.push(junction.last);
    }
}

impl Widget for Table {
    fn render(&self, writer: &mut WidgetWriter) {
        let TableState::HasContent { cells_count, cells_len, content } = &self.state else {
            panic!("trying to render an empty table");
        };

        if let Some(title) = &self.title {
            writer.push_str("  ");
            writer.push_str(title);
            writer.push_str("\n");
        }
        self.render_horizontal_border(writer, cells_len, &self.charset.first_junction);
        writer.push_str("\n");

        let mut rows = content.chunks_exact(*cells_count);
        let mut idx = 0;
        for row in rows.by_ref() {
            let last_row = idx == content.len() / cells_count - 1;
            let lines_count = row.iter().map(|cell| cell.len()).max().unwrap_or(0);

            for line in 0..lines_count {
                writer.push(self.charset.vertical_separator);
                for (idx, cell_all_lines) in row.iter().enumerate() {
                    let cell = cell_all_lines.get(line).map(|c| c.as_str()).unwrap_or_default();
                    writer.push(' ');
                    writer.push_str(cell);

                    // Padding to align all cells.
                    for _ in cell.len()..cells_len[idx] {
                        writer.push(' ');
                    }

                    writer.push(' ');
                    writer.push(self.charset.vertical_separator);
                }
                writer.push('\n');
            }
            self.render_horizontal_border(
                writer,
                cells_len,
                if last_row { &self.charset.last_junction } else { &self.charset.middle_junction },
            );
            if !last_row {
                writer.push('\n');
            }

            idx += 1;
        }

        assert!(rows.remainder().is_empty());
    }
}

enum TableState {
    Empty,
    HasContent { cells_count: usize, cells_len: Vec<usize>, content: Vec<Vec<String>> },
}

struct TableCharset {
    vertical_separator: char,
    horizontal_separator: char,
    first_junction: TableJunctionCharset,
    middle_junction: TableJunctionCharset,
    last_junction: TableJunctionCharset,
}

struct TableJunctionCharset {
    first: char,
    middle: char,
    last: char,
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::configure_insta;
    use insta::assert_snapshot;

    #[test]
    fn test_sample_table() {
        let _config = configure_insta();

        let mut table = Table::new();
        table.add_row(["Foo", "Bar", "Baz"]);
        table.add_row(["Hello", "super long", "world!"]);
        table.add_row(["98%", "", "-"]);

        assert_snapshot!(table.render_to_string());
    }

    #[test]
    fn test_single_cell_table() {
        let _config = configure_insta();

        let mut table = Table::new();
        table.add_row(["alone"]);

        assert_snapshot!(table.render_to_string());
    }

    #[test]
    fn test_table_with_title() {
        let _config = configure_insta();

        let mut table = Table::new();
        table.set_title("Example title:");
        table.add_row(["a", "b", "c"]);

        assert_snapshot!(table.render_to_string());
    }

    #[test]
    fn test_table_with_multiple_lines() {
        let _config = configure_insta();

        let mut table = Table::new();
        table.add_row(["a", "b", "c"]);
        table.add_row(["foo\nbar", "baz", "qu\nu\n\n\nx!!!!!!!!"]);

        assert_snapshot!(table.render_to_string());
    }

    #[test]
    #[should_panic = "other rows have 3 cells, while this row has 2 cells"]
    fn test_wrong_number_of_cells() {
        let mut table = Table::new();
        table.add_row(["foo", "bar", "baz"]);
        table.add_row(["a", "b"]);
    }

    #[test]
    #[should_panic = "trying to render an empty table"]
    fn test_empty_table() {
        Table::new().render_to_string();
    }
}
