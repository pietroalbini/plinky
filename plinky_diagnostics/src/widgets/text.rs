use crate::widgets::Widget;
use crate::WidgetWriter;

pub struct Text {
    content: String,
}

impl Text {
    pub fn new(content: impl Into<String>) -> Self {
        Self { content: content.into() }
    }
}

impl Widget for Text {
    fn render(&self, writer: &mut WidgetWriter) {
        writer.push_str(&self.content);
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_render_text() {
        assert_eq!("Hello world!", Text::new("Hello world!").render_to_string());
    }
}
