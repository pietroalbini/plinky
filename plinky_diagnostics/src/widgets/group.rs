use crate::widgets::Widget;
use crate::WidgetWriter;
use crate::writer::IndentMode;

pub struct WidgetGroup {
    widgets: Vec<Box<dyn Widget>>,
}

impl WidgetGroup {
    pub fn new() -> Self {
        WidgetGroup { widgets: Vec::new() }
    }

    pub fn add<T: Widget + 'static>(mut self, widget: T) -> Self {
        self.widgets.push(Box::new(widget));
        self
    }
}

impl Widget for WidgetGroup {
    fn render(&self, writer: &mut WidgetWriter) {
        writer.push_indent(" │", IndentMode::ShowAlways);
        writer.push_indent("  ", IndentMode::HideOnEmptyLines);
        for (idx, widget) in self.widgets.iter().enumerate() {
            if idx == 0 {
                writer.push_str("\n");
            } else {
                writer.push_str("\n\n");
            }
            widget.render(writer);
            if writer.last_char() == Some('\n') {
                panic!("widgets must not terminate with a newline");
            }
        }
        writer.pop_indent();
        writer.pop_indent();
        if !self.widgets.is_empty() {
            writer.push_str("\n ┴");
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::configure_insta;
    use crate::widgets::{Table, Text};
    use insta::assert_snapshot;

    #[test]
    fn test_empty_group() {
        assert!(WidgetGroup::new().render_to_string().is_empty());
    }

    #[test]
    fn test_multiple_widgets() {
        let _config = configure_insta();

        let mut table = Table::new();
        table.add_row(["Foo", "Bar"]);

        let group = WidgetGroup::new().add(Text::new("A simple text message!")).add(table);
        assert_snapshot!(group.render_to_string());
    }
}
