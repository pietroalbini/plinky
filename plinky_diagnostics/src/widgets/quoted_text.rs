use crate::WidgetWriter;
use crate::widgets::Widget;
use crate::writer::IndentMode;

pub struct QuotedText {
    content: String,
}

impl QuotedText {
    pub fn new(content: impl Into<String>) -> Self {
        Self { content: content.into() }
    }
}

impl Widget for QuotedText {
    fn render(&self, writer: &mut WidgetWriter) {
        if self.content.is_empty() {
            return;
        }

        writer.push_str("╭\n");

        writer.push_indent("│", IndentMode::ShowAlways);
        writer.push_indent(" ", IndentMode::HideOnEmptyLines);

        writer.push_str(&self.content);
        writer.push('\n');

        writer.pop_indent();
        writer.pop_indent();

        writer.push_str("╰");
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::configure_insta;
    use insta::assert_snapshot;

    #[test]
    fn test_empty_content() {
        assert_eq!("", QuotedText::new("").render_to_string());
    }

    #[test]
    fn test_single_line() {
        let _config = configure_insta();

        let content = QuotedText::new("Hello world");
        assert_snapshot!(content.render_to_string());
    }

    #[test]
    fn test_multiple_lines() {
        let _config = configure_insta();

        let content = QuotedText::new("Hello world\nThis\n\n  has\n    multiple lines!");
        assert_snapshot!(content.render_to_string());
    }
}
