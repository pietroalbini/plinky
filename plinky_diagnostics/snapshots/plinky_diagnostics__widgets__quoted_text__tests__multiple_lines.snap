---
source: plinky_diagnostics/src/widgets/quoted_text.rs
expression: content.render_to_string()
---
╭
│ Hello world
│ This
│
│   has
│     multiple lines!
╰
